File: DSA/LeetCode/Arrays/1365. How Many Numbers Are Smaller Than the Current Number/Solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var smallerNumbersThanCurrent = function (nums) {
  // Brute force solution
  /* result = []
  for(let i=0;i<nums.length;i++) {
    let lessElements = 0
    
    for(let j = 0;j<nums.length;j++) {
      if(nums[i] != nums[j] && nums[i] > nums[j]) {
        lessElements++;
      }
    }
    result.push(lessElements)
  }
  return result*/

  // Optimal approach
  // Sort the array in descending order
  result = [...nums].sort((a, b) => b - a)
  hashMap = {}

  // Create a map of elements with the number of elements less than it
  const countMap = new Map(result.map((ele, index) => [ele, result.length - index - 1]));

  return nums.map((ele) => countMap.get(ele))
};

File: DSA/LeetCode/Arrays/1365. How Many Numbers Are Smaller Than the Current Number/Solution.py
class Solution:
    def smallerNumbersThanCurrent(self, nums: List[int]) -> List[int]:
        # Brute force solution
        '''
        result = []
        length_nums = len(nums)

        for i in range(length_nums):
          numbers = 0

          for j in range(length_nums):
            if i != j and nums[i] > nums[j]:
              numbers += 1

          result.append(numbers)

        return result
        '''

        # Optimal solution
        len_nums = len(nums)
        sorted_nums = nums[:]
        sorted_nums.sort()

        hashMap = {}
        for i in range(len_nums):
            if sorted_nums[i] not in hashMap:
                hashMap[sorted_nums[i]] = i

        for i in range(len_nums):
            nums[i] = hashMap[nums[i]]

        return nums
